File: src/conversation_search/core/summarization.py
#!/usr/bin/env python3
"""
Smart content extraction for conversation-search
Uses hybrid approach: full user content + smart extraction for assistant messages
"""

import json
import re
import sqlite3
import sys
from pathlib import Path
from typing import List, Dict, Optional, Tuple

from conversation_search.core.db import connect as connect_db


class MessageSummarizer:
    """Handles smart hybrid extraction without AI summarization"""

    def __init__(self, db_path: str = "~/.conversation-search/index.db"):
        self.db_path = Path(db_path).expanduser()

    def is_tool_noise(self, message: Dict) -> bool:
        """
        Detect if message is pure tool spam that should be filtered

        Tool noise characteristics:
        - Pure tool markers with minimal text
        - Common read/search operations
        - Very short assistant acknowledgments
        """
        content = message['content']
        msg_type = message['message_type']

        # Pure tool noise - ONLY tool markers, no substantial content
        # Don't use this check, rely on the pattern matching below instead

        # Tool results are always noise
        if content.strip() == '[Tool result]':
            return True

        # Request interrupted messages
        if '[Request interrupted' in content:
            return True

        # Empty or whitespace-only content
        if not content.strip():
            return True

        # Very short messages that are just tool markers
        if len(content) < 50:
            return False  # These get marked as "too_short" instead

        # Common noise patterns with substantial text check
        noise_patterns = [
            '[Tool: Read]',
            '[Tool: Glob]',
            '[Tool: LS]',
            '[Tool: Grep]',
            '[Tool result]',
            '[Request interrupted]',
        ]

        if any(pattern in content for pattern in noise_patterns):
            # But allow if there's substantial text overall
            # Remove all tool markers and check remaining text
            text_without_tools = content
            for pattern in noise_patterns:
                text_without_tools = text_without_tools.replace(pattern, '')
            text_without_tools = text_without_tools.strip()

            if len(text_without_tools) > 100:
                return False
            return True

        # Assistant messages that are just acknowledging tool use
        if msg_type == 'assistant' and len(content) < 150:
            if any(phrase in content.lower() for phrase in [
                'let me read', 'let me check', 'let me search',
                "i'll look at", 'looking at', 'checking'
            ]):
                return True

        return False

    def get_searchable_text(self, message: Dict) -> str:
        """
        Extract searchable text using smart hybrid extraction

        Strategy:
        - User messages: index full content (they're concise, avg 3.5K chars)
        - Assistant messages: first 500 + last 200 chars + tool usage
        """
        content = message['content']
        msg_type = message['message_type']

        # User messages: usually concise, index full content
        if msg_type == 'user':
            return content  # Avg 3.5K chars, important info upfront

        # Assistant messages: verbose, extract strategically
        lines = content.split('\n')

        # Take first 500 chars
        first_part = content[:500] if len(content) > 500 else content

        # Extract tool mentions (important markers)
        tools = re.findall(r'\[Tool:\s*(\w+)\]', content)
        tool_summary = f"\nTools used: {', '.join(set(tools))}" if tools else ""

        # Last 200 chars often have conclusion
        last_part = ""
        if len(content) > 700:
            last_part = f"\n...\n{content[-200:]}"

        return f"{first_part}{tool_summary}{last_part}"

    def needs_summarization(self, message: Dict) -> Tuple[bool, str]:
        """
        Check if message needs processing (no AI, just classification)

        Returns: (needs_processing, reason)

        Reasons:
        - 'tool_noise': Pure tool spam, mark as noise
        - 'too_short': Message is < 50 chars
        - 'extract': Needs smart extraction (always for non-noise messages)
        """
        # Tool noise (check first, regardless of length)
        if self.is_tool_noise(message):
            return False, 'tool_noise'

        # Too short
        if len(message['content']) < 50:
            return False, 'too_short'

        # Everything else gets smart extraction
        return True, 'extract'

    def extract_batch(self, messages: List[Dict]) -> List[Dict]:
        """
        Extract searchable text from messages using smart hybrid extraction

        Returns: List of dicts with uuid, summary (extracted text)
        """
        if not messages:
            return []

        extractions = []
        for message in messages:
            extracted_text = self.get_searchable_text(message)
            extractions.append({
                'uuid': message['uuid'],
                'summary': extracted_text,
                'message_type': message['message_type']
            })

        return extractions

    def update_database(self, summaries: List[Dict], method: str = 'smart_extraction'):
        """Update database with extracted searchable text"""
        conn = connect_db(str(self.db_path))
        cursor = conn.cursor()

        updated = 0
        try:
            for summary_data in summaries:
                uuid = summary_data.get('uuid')
                summary = summary_data.get('summary')

                if not uuid or not summary:
                    continue

                cursor.execute("""
                    UPDATE messages
                    SET summary = ?, is_summarized = TRUE, summary_method = ?
                    WHERE message_uuid = ?
                """, (summary, method, uuid))

                if cursor.rowcount > 0:
                    updated += 1

            conn.commit()
        except sqlite3.Error as e:
            conn.rollback()
            print(f"Error updating summaries: {e}", file=sys.stderr)
            raise
        finally:
            conn.close()

        return updated

    def mark_tool_noise(self, message_uuids: List[str]):
        """Mark messages as tool noise in database"""
        if not message_uuids:
            return

        conn = connect_db(str(self.db_path))
        cursor = conn.cursor()

        try:
            placeholders = ','.join('?' * len(message_uuids))
            cursor.execute(f"""
                UPDATE messages
                SET is_tool_noise = TRUE, summary_method = 'too_short'
                WHERE message_uuid IN ({placeholders})
            """, message_uuids)
            conn.commit()
        except sqlite3.Error as e:
            conn.rollback()
            print(f"Error marking tool noise: {e}", file=sys.stderr)
            raise
        finally:
            conn.close()

    def mark_too_short(self, message_uuids: List[str]):
        """Mark messages as too short to need summarization"""
        if not message_uuids:
            return

        conn = connect_db(str(self.db_path))
        cursor = conn.cursor()

        try:
            placeholders = ','.join('?' * len(message_uuids))
            cursor.execute(f"""
                UPDATE messages
                SET is_summarized = TRUE, summary_method = 'too_short'
                WHERE message_uuid IN ({placeholders})
            """, message_uuids)
            conn.commit()
        except sqlite3.Error as e:
            conn.rollback()
            print(f"Error marking too short: {e}", file=sys.stderr)
            raise
        finally:
            conn.close()


def message_uses_conversation_search(message: Dict) -> bool:
    """
    Detect if a message involves using the conversation-search tool.

    Returns True if this is a Claude message that used the search tool.
    This helps identify meta-conversations that should be marked.

    Args:
        message: Message dict with 'message_type' and 'content'

    Returns:
        True if message uses conversation-search, False otherwise
    """
    # Only assistant messages can use tools
    if message.get('message_type') != 'assistant':
        return False

    content = message.get('content', '')
    content_lower = content.lower()

    # Pattern 1: Bash tool + ai-conversation-search command
    # This means Claude RAN the command via Bash
    if '[Tool: Bash]' in content and 'ai-conversation-search' in content:
        return True

    # Pattern 2: Direct ai-conversation-search command usage
    # Multiple patterns to catch various invocation styles
    if 'ai-conversation-search' in content:
        cmd_patterns = [
            # Direct subcommand: ai-conversation-search search
            r'ai-conversation-search\s+(search|list|index|tree|context|resume)',
            # Flags before subcommand: ai-conversation-search --json search
            r'ai-conversation-search\s+--\w+\s+(search|list|index|tree|context|resume)',
            # Version/help flags: ai-conversation-search --help
            r'ai-conversation-search\s+(--help|--version|-h|-v)',
        ]
        for pattern in cmd_patterns:
            if re.search(pattern, content):
                return True

    # Pattern 3: Tool upgrade commands
    if re.search(r'uv\s+tool\s+upgrade\s+ai-conversation-search', content):
        return True
    if re.search(r'pip\s+install\s+--upgrade\s+ai-conversation-search', content):
        return True

    # Pattern 4: Command existence checks
    if re.search(r'command\s+-v\s+ai-conversation-search', content):
        return True
    if re.search(r'which\s+ai-conversation-search', content):
        return True

    # Pattern 5: Skill activation markers (actual usage, not discussion)
    if 'conversation-search skill is loading' in content_lower:
        return True
    if 'conversation-search skill is running' in content_lower:
        return True

    # Pattern 6: The quoted skill name with activation verbs (more specific than before)
    if '"conversation-search"' in content_lower:
        activation_verbs = ['loading', 'running', 'is active', 'activated']
        if any(verb in content_lower for verb in activation_verbs):
            return True

    # Pattern 7: Skill allowed tools marker (with proximity check)
    if 'allowed 1 tools for this command' in content_lower:
        marker_pos = content_lower.find('allowed 1 tools')
        search_pos = content_lower.find('conversation-search')
        if search_pos != -1 and abs(marker_pos - search_pos) < 100:
            return True

    return False


def is_summarizer_conversation(conv_file: Path, messages: List[Dict]) -> bool:
    """
    Detect if this is an automated summarizer conversation

    Characteristics:
    - Very short (2-5 messages)
    - Contains summarization keywords
    - No tool use complexity
    """
    # Wrong length for summarizer
    if len(messages) < 2 or len(messages) > 10:
        return False

    # Check first user message for summarization patterns
    first_user = next((m for m in messages if m['message_type'] == 'user'), None)
    if not first_user:
        return False

    content = first_user['content'].lower()

    # Summarization keywords
    indicators = [
        'summarize this',
        'create a 1-2 sentence summary',
        'generate concise summaries',
        'max 150 characters',
        'for each message',
        'json output:',
        'brief summary here',
        'messages to summarize:',
    ]

    return any(indicator in content for indicator in indicators)

File: src/summarization.rs
use regex::Regex;
use std::sync::LazyLock;

static CMD_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    vec![
        Regex::new(r"ai-conversation-search\s+(search|list|index|tree|context|resume)").unwrap(),
        Regex::new(r"ai-conversation-search\s+--\w+\s+(search|list|index|tree|context|resume)").unwrap(),
        Regex::new(r"ai-conversation-search\s+(--help|--version|-h|-v)").unwrap(),
        Regex::new(r"uv\s+tool\s+upgrade\s+ai-conversation-search").unwrap(),
        Regex::new(r"pip\s+install\s+--upgrade\s+ai-conversation-search").unwrap(),
        Regex::new(r"command\s+-v\s+ai-conversation-search").unwrap(),
        Regex::new(r"which\s+ai-conversation-search").unwrap(),
    ]
});

const NOISE_PATTERNS: &[&str] = &[
    "[Tool: Read]",
    "[Tool: Glob]",
    "[Tool: LS]",
    "[Tool: Grep]",
    "[Tool result]",
    "[Request interrupted]",
];

const ACK_PHRASES: &[&str] = &[
    "let me read",
    "let me check",
    "let me search",
    "i'll look at",
    "looking at",
    "checking",
];

/// Detect if a message is pure tool spam that should be filtered.
pub fn is_tool_noise(content: &str, message_type: &str) -> bool {
    let trimmed = content.trim();

    if trimmed == "[Tool result]" {
        return true;
    }
    if trimmed.contains("[Request interrupted") {
        return true;
    }
    if trimmed.is_empty() {
        return true;
    }
    if content.len() < 50 {
        return false; // marked as "too_short" instead
    }

    if NOISE_PATTERNS.iter().any(|p| content.contains(p)) {
        let mut text_without_tools = content.to_string();
        for pattern in NOISE_PATTERNS {
            text_without_tools = text_without_tools.replace(pattern, "");
        }
        let remaining = text_without_tools.trim();
        if remaining.len() > 100 {
            return false;
        }
        return true;
    }

    if message_type == "assistant" && content.len() < 150 {
        let lower = content.to_lowercase();
        if ACK_PHRASES.iter().any(|phrase| lower.contains(phrase)) {
            return true;
        }
    }

    false
}

/// Detect if a message involves using the conversation-search tool.
pub fn message_uses_conversation_search(content: &str, message_type: &str) -> bool {
    if message_type != "assistant" {
        return false;
    }

    // Pattern 1: Bash tool + ai-conversation-search command
    if content.contains("[Tool: Bash]") && content.contains("ai-conversation-search") {
        return true;
    }

    // Pattern 2: Direct command usage
    if content.contains("ai-conversation-search") {
        for re in CMD_PATTERNS.iter() {
            if re.is_match(content) {
                return true;
            }
        }
    }

    let content_lower = content.to_lowercase();

    // Pattern 5: Skill activation markers
    if content_lower.contains("conversation-search skill is loading")
        || content_lower.contains("conversation-search skill is running")
    {
        return true;
    }

    // Pattern 6: Quoted skill name with activation verbs
    if content_lower.contains("\"conversation-search\"") {
        let activation_verbs = ["loading", "running", "is active", "activated"];
        if activation_verbs.iter().any(|v| content_lower.contains(v)) {
            return true;
        }
    }

    // Pattern 7: Skill allowed tools marker
    if content_lower.contains("allowed 1 tools for this command") {
        if let Some(marker_pos) = content_lower.find("allowed 1 tools") {
            if let Some(search_pos) = content_lower.find("conversation-search") {
                let diff = if marker_pos > search_pos {
                    marker_pos - search_pos
                } else {
                    search_pos - marker_pos
                };
                if diff < 100 {
                    return true;
                }
            }
        }
    }

    false
}

/// Detect if this is an automated summarizer conversation.
pub fn is_summarizer_conversation(messages: &[crate::indexer::Message]) -> bool {
    if messages.len() < 2 || messages.len() > 10 {
        return false;
    }

    let first_user = messages.iter().find(|m| m.message_type == "user");
    let first_user = match first_user {
        Some(m) => m,
        None => return false,
    };

    let content = first_user.content.to_lowercase();
    let indicators = [
        "summarize this",
        "create a 1-2 sentence summary",
        "generate concise summaries",
        "max 150 characters",
        "for each message",
        "json output:",
        "brief summary here",
        "messages to summarize:",
    ];

    indicators.iter().any(|ind| content.contains(ind))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tool_noise_tool_result() {
        assert!(is_tool_noise("[Tool result]", "assistant"));
    }

    #[test]
    fn test_tool_noise_empty() {
        assert!(is_tool_noise("", "user"));
        assert!(is_tool_noise("   ", "user"));
    }

    #[test]
    fn test_tool_noise_short_not_noise() {
        assert!(!is_tool_noise("Hello world", "user"));
    }

    #[test]
    fn test_tool_noise_with_substantial_text() {
        let content = format!(
            "[Tool: Read] {} some meaningful content that is more than 100 characters long and should not be filtered as noise because it contains real information",
            "x".repeat(50)
        );
        assert!(!is_tool_noise(&content, "assistant"));
    }

    #[test]
    fn test_message_uses_conversation_search() {
        assert!(message_uses_conversation_search(
            "[Tool: Bash] ai-conversation-search search test",
            "assistant"
        ));
        assert!(!message_uses_conversation_search(
            "ai-conversation-search search test",
            "user"
        ));
    }

    // --- is_tool_noise additional tests ---

    #[test]
    fn test_tool_noise_request_interrupted() {
        assert!(is_tool_noise("[Request interrupted by user]", "user"));
    }

    #[test]
    fn test_tool_noise_with_tools_but_short_remaining() {
        let content = "[Tool: Read] some short text here that is over 50 chars padding padding";
        assert!(is_tool_noise(content, "assistant"));
    }

    #[test]
    fn test_tool_noise_ack_phrase_let_me_check() {
        // Must be >= 50 chars and < 150 chars to hit the ack phrase branch
        let content = "Let me check the file for you, I will look into it right away now";
        assert!(content.len() >= 50 && content.len() < 150);
        assert!(is_tool_noise(content, "assistant"));
    }

    #[test]
    fn test_tool_noise_ack_phrase_wrong_type() {
        let content = "Let me check the file for you, I will look into it right away now";
        assert!(!is_tool_noise(content, "user"));
    }

    #[test]
    fn test_tool_noise_long_ack_not_noise() {
        let content = format!(
            "Let me check the file for you. {}",
            "This is a very detailed explanation that goes on and on. ".repeat(5)
        );
        assert!(content.len() > 150);
        assert!(!is_tool_noise(&content, "assistant"));
    }

    // --- message_uses_conversation_search additional tests ---

    #[test]
    fn test_search_skill_loading_marker() {
        assert!(message_uses_conversation_search(
            "The conversation-search skill is loading now",
            "assistant"
        ));
    }

    #[test]
    fn test_search_skill_running_marker() {
        assert!(message_uses_conversation_search(
            "The conversation-search skill is running",
            "assistant"
        ));
    }

    #[test]
    fn test_search_quoted_skill_with_verb() {
        assert!(message_uses_conversation_search(
            "The \"conversation-search\" plugin has been activated successfully",
            "assistant"
        ));
    }

    #[test]
    fn test_search_allowed_tools_nearby() {
        let content = "conversation-search: allowed 1 tools for this command";
        assert!(message_uses_conversation_search(content, "assistant"));
    }

    #[test]
    fn test_search_allowed_tools_far() {
        let padding = "x".repeat(120);
        let content = format!(
            "conversation-search {}allowed 1 tools for this command",
            padding
        );
        assert!(!message_uses_conversation_search(&content, "assistant"));
    }

    #[test]
    fn test_search_user_message_ignored() {
        assert!(!message_uses_conversation_search(
            "conversation-search skill is loading and allowed 1 tools for this command",
            "user"
        ));
    }

    // --- is_summarizer_conversation tests ---

    use crate::indexer::Message;

    fn make_message(msg_type: &str, content: &str) -> Message {
        Message {
            uuid: "test-uuid".to_string(),
            parent_uuid: None,
            is_sidechain: false,
            timestamp: Some("2025-01-15T10:00:00Z".to_string()),
            message_type: msg_type.to_string(),
            content: content.to_string(),
            session_id: Some("test-session".to_string()),
            is_meta_conversation: false,
        }
    }

    #[test]
    fn test_summarizer_with_summarize_this() {
        let messages = vec![
            make_message("user", "Please summarize this conversation for me"),
            make_message("assistant", "Here is the summary..."),
        ];
        assert!(is_summarizer_conversation(&messages));
    }

    #[test]
    fn test_summarizer_with_json_output() {
        let messages = vec![
            make_message("user", "Process these messages. json output: required"),
            make_message("assistant", "Done."),
        ];
        assert!(is_summarizer_conversation(&messages));
    }

    #[test]
    fn test_summarizer_too_few_messages() {
        let messages = vec![make_message("user", "summarize this")];
        assert!(!is_summarizer_conversation(&messages));
    }

    #[test]
    fn test_summarizer_too_many_messages() {
        let messages: Vec<Message> = (0..11)
            .map(|i| {
                if i == 0 {
                    make_message("user", "summarize this")
                } else if i % 2 == 0 {
                    make_message("user", "more")
                } else {
                    make_message("assistant", "ok")
                }
            })
            .collect();
        assert!(!is_summarizer_conversation(&messages));
    }

    #[test]
    fn test_summarizer_no_user_message() {
        let messages = vec![
            make_message("assistant", "summarize this"),
            make_message("assistant", "ok"),
        ];
        assert!(!is_summarizer_conversation(&messages));
    }

    #[test]
    fn test_summarizer_normal_conversation() {
        let messages = vec![
            make_message("user", "How do I use Rust iterators?"),
            make_message("assistant", "You can use .iter() to create an iterator..."),
            make_message("user", "Thanks!"),
        ];
        assert!(!is_summarizer_conversation(&messages));
    }

    // --- proximity boundary tests ---

    #[test]
    fn test_search_proximity_exact_boundary() {
        // "conversation-search" is 19 chars; we need "allowed 1 tools" to start within 100 chars
        // of "conversation-search". Place padding so the distance is exactly 99.
        let padding = "x".repeat(80); // 19 + 80 = 99 distance
        let content = format!(
            "conversation-search{}allowed 1 tools for this command",
            padding
        );
        assert!(message_uses_conversation_search(&content, "assistant"));
    }

    #[test]
    fn test_search_proximity_over_boundary() {
        // Make distance > 100
        let padding = "x".repeat(101);
        let content = format!(
            "conversation-search{}allowed 1 tools for this command",
            padding
        );
        assert!(!message_uses_conversation_search(&content, "assistant"));
    }
}
